From 7c1e5fae95974502acf21f56da376c9d99fdbc28 Mon Sep 17 00:00:00 2001
From: Vaibhav Hiremath <hvaibhav@ti.com>
Date: Tue, 24 Feb 2009 11:58:33 +0530
Subject: [PATCH 317/426] SDOCM00053743: kernel crash/hang issue on wrong rotation value fixed

Kernel used hand/crash for the wrong value passed for
var.rotate (other than 0, 1, 2, 3), fixxed under this commit.

Now the driver checks for the wring rotation value and returns error.

Only part of IR fixed under this commit.
---
 drivers/video/omap2/omapfb-main.c |   12 ++++++++++--
 1 files changed, 10 insertions(+), 2 deletions(-)

diff --git a/drivers/video/omap2/omapfb-main.c b/drivers/video/omap2/omapfb-main.c
index bdd9926..2da043f 100644
--- a/drivers/video/omap2/omapfb-main.c
+++ b/drivers/video/omap2/omapfb-main.c
@@ -35,8 +35,6 @@
 
 #define MODULE_NAME     "omapfb"
 
-#define omap_rotation_index(deg) (deg/90)
-
 static char *def_mode;
 static char *def_vram;
 static int def_rotate_type;
@@ -44,6 +42,12 @@ static int def_rotate = -1;
 
 #define VRFB_WIDTH 	(2048)
 #define FB_SIZE 	(1280 * 720 * 4 * 2)
+
+#define IS_VALID_ROTATION(rotate) ((rotate == FB_ROTATE_UR) || \
+				(rotate == FB_ROTATE_CW) || \
+				(rotate == FB_ROTATE_UD) || \
+				(rotate == FB_ROTATE_CCW))
+
 #ifdef DEBUG
 unsigned int omapfb_debug;
 module_param_named(debug, omapfb_debug, bool, 0644);
@@ -401,6 +405,10 @@ int check_fb_var(struct fb_info *fbi, struct fb_var_screeninfo *var)
 	if (var->rotate != fbi->var.rotate) {
 		DBG("rotation changing\n");
 
+		if (!IS_VALID_ROTATION(var->rotate)) {
+			DBG("invalid rotation parameter\n");
+			return -EINVAL;
+		}
 		ofbi->rotation = var->rotate;
 
 		if (abs(var->rotate - fbi->var.rotate) != 2) {
-- 
1.6.2.4

